Release Standard Operating Procedures (SOP) (#3100)
grafana/agent: docs/developer/release/1-create-release-branch.md

# Create Release Branch

A single release branch is created for every major or minor release. That release
branch is then used for all Release Candidates, the Stable Release, and all
Patch Releases.

> ## Release Version Prefix

Reviewer: IMO we should be talking about this in terms of semantic versioning rather than the first/second/third numbers, but I think this section is OK for now.

> The `[release version prefix]` can be determined by looking at the last release

Reviewer: ```suggestion
The `RELEASE_VERSION_PREFIX` can be determined by looking at the last release
```

> version and adding to it. Also, drop off the 3rd number as shown in the examples below.

Reviewer: Spell out ordinals less than 10, unless they're used for math:

```suggestion
version and adding 1 to it. Also, drop off the third number as shown in the examples below:
```

From Google's developer guide on [Numbers](https://developers.google.com/style/numbers).

- v0.30.0 -> v0.31
- v0.30.3 -> v0.31

> For a major version we jump up the first number as shown in the examples below.

Reviewer: ```suggestion
For a major version we jump up the first number as shown in the examples below:
```

- v0.10.0 -> v1.0
- v0.31.0 -> v1.0

> *NOTE: This value will be referred to as `[release version prefix]` in this documentation*

Reviewer: ```suggestion
*NOTE: This value will be referred to as `RELEASE_VERSION_PREFIX` in this documentation*
```

## Steps

1. Determine which commit should be used as a base for the release branch.

2. Create and push the release branch from the selected base commit:

>     The name of the release branch should be `release-[release version prefix]`

Reviewer: ```suggestion
    The name of the release branch should be `release-RELEASE_VERSION_PREFIX`
```

    defined above, such as `release-v0.31`.

>         *NOTE: There is never a branch such as `release-v0.31-rc.0` or `release-v0.31.0`.*

Reviewer: ```suggestion
        > **NOTE**: Branches are only made for RELEASE_VERSION_PREFIX; do not create branches for the full RELEASE_VERSION such as `release-v0.31-rc.0` or `release-v0.31.0`.
```

    - If the consensus commit is the latest commit from main you can branch from main.
    - If the consensus commit is not the latest commit from main, branch from that instead.

>     *NOTE: don't create any other branches that are prefixed with `release` when creating PRs or
>     those branches will collide with our automated release build publish rules.*

Reviewer: ```suggestion
    > **NOTE**: Don't create any other branches that are prefixed with `release` when creating PRs or
    those branches will collide with our automated release build publish rules.
```

# Create Release Branch

A single release branch is created for every major or minor release. That release
branch is then used for all Release Candidates, the Stable Release, and all
Patch Releases.

## Release Version Prefix

The `[release version prefix]` can be determined by looking at the last release
version and adding to it. Also, drop off the 3rd number as shown in the examples below.

- v0.30.0 -> v0.31
- v0.30.3 -> v0.31

For a major version we jump up the first number as shown in the examples below.

- v0.10.0 -> v1.0
- v0.31.0 -> v1.0

*NOTE: This value will be referred to as `[release version prefix]` in this documentation*

## Steps

> 1. Gather consensus on which commit should be used as a base for the release

Reviewer 2: I would change this to `Determine which commit` IMO the release shepherd has control and doesnt necessarily need consensus.
